{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'sbmlcore'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[1;32m/Users/dylanadlard/Documents/Oxford/Year_5/SABS/Rotations/Rotation_1_report/SABS_rotation_1_report/Report.ipynb Cell 2'\u001b[0m in \u001b[0;36m<cell line: 7>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/dylanadlard/Documents/Oxford/Year_5/SABS/Rotations/Rotation_1_report/SABS_rotation_1_report/Report.ipynb#ch0000001?line=4'>5</a>\u001b[0m sys\u001b[39m.\u001b[39mpath\u001b[39m.\u001b[39minsert(\u001b[39m0\u001b[39m, \u001b[39m'\u001b[39m\u001b[39m../sbmlcore\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/dylanadlard/Documents/Oxford/Year_5/SABS/Rotations/Rotation_1_report/SABS_rotation_1_report/Report.ipynb#ch0000001?line=5'>6</a>\u001b[0m sys\u001b[39m.\u001b[39mpath\u001b[39m.\u001b[39minsert(\u001b[39m0\u001b[39m, \u001b[39m'\u001b[39m\u001b[39m/Users/dylanadlard/Library/\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/dylanadlard/Documents/Oxford/Year_5/SABS/Rotations/Rotation_1_report/SABS_rotation_1_report/Report.ipynb#ch0000001?line=6'>7</a>\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39msbmlcore\u001b[39;00m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'sbmlcore'"
     ]
    }
   ],
   "source": [
    "import ModelSearch\n",
    "import pandas as pd \n",
    "import sys\n",
    "\n",
    "sys.path.insert(0, '../sbmlcore')\n",
    "sys.path.insert(0, '/Users/dylanadlard/Library/')\n",
    "import sbmlcore from \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Hello[WHO2021]<a id='WHO2021'></a> Goodbye"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[Blurf](#WHO2021)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Methods and Results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data Processing Workflow"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The pipeline below (FIGURE) first trains models on binary phenotype readouts (resistance vs susceptible) and leads to a prediction of a binary phenotype. This pipeline requires the pulldown of structural, biochemical, thermodynamic, and evolutionary features from RNAP and its mutations before merging these with filtered, prepated mutation data. These data and that for the filtered, prepared phonetype preadouts form the feature and target sets, respectively. \n",
    "\n",
    "The data is split into a 70:30 training and test set, the former of which undergoes a parameter grid search with cross validation and upsampling (discussed later). The model is retrained using the best parameters, and a classification report and confusion matrix is outputted post decision-threshold shifting to maximise specificity."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![title](images/direct_binary_class_workflow.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Target Set Derivation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mutation data derivation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The data that formed the target and much of the training sets were derived from the CRyPTIC database. Mutation data was extracted from Illumina platform WGS of clinical isolates. Since this study focuses on predicting susceptiblity in response to solo structural SNPs in Mtb RNA polymerase, mutations associated with other drug targets, non-solo mutations, phylogenetic, and synonymous mutaations were filtered out, as well as insertion/deletion polymorphisms (indels), and mutations in the promoter. \n",
    "\n",
    "Of particular note is the fact mutations in the RRDR are reportedly responsible for more than 95% of resistance to rifampicin ([Mvelase2019](#Mvelase2019)); a statistic that is at least partially reflected in our data. Therefore, as this is a preliminary investigation, and for simplicity, only mutations located in rpoB are initially considered."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>AMINO_ACID_NUMBER</th>\n",
       "      <th>GENOME_INDEX</th>\n",
       "      <th>NUCLEOTIDE_NUMBER</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IS_INDEL</th>\n",
       "      <th>...</th>\n",
       "      <th>IS_NULL</th>\n",
       "      <th>IS_FILTER_PASS</th>\n",
       "      <th>ELEMENT_TYPE</th>\n",
       "      <th>MUTATION_TYPE</th>\n",
       "      <th>INDEL_LENGTH</th>\n",
       "      <th>INDEL_1</th>\n",
       "      <th>INDEL_2</th>\n",
       "      <th>SITEID</th>\n",
       "      <th>NUMBER_NUCLEOTIDE_CHANGES</th>\n",
       "      <th>segid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>site.02.subj.1033.lab.2013185075.iso.1</th>\n",
       "      <td>rpoB</td>\n",
       "      <td>S450L</td>\n",
       "      <td>450.0</td>\n",
       "      <td>450.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tcg</td>\n",
       "      <td>ttg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>02</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.02.subj.0739.lab.2013221518.iso.1</th>\n",
       "      <td>rpoB</td>\n",
       "      <td>S450L</td>\n",
       "      <td>450.0</td>\n",
       "      <td>450.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tcg</td>\n",
       "      <td>ttg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>02</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.02.subj.0604.lab.241032-14.iso.1</th>\n",
       "      <td>rpoB</td>\n",
       "      <td>M587I</td>\n",
       "      <td>587.0</td>\n",
       "      <td>587.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>atg</td>\n",
       "      <td>ata</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>02</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.02.subj.0104.lab.22A057.iso.1</th>\n",
       "      <td>rpoB</td>\n",
       "      <td>S450L</td>\n",
       "      <td>450.0</td>\n",
       "      <td>450.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tcg</td>\n",
       "      <td>ttg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>02</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.02.subj.0885.lab.22A119.iso.1</th>\n",
       "      <td>rpoB</td>\n",
       "      <td>S450L</td>\n",
       "      <td>450.0</td>\n",
       "      <td>450.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tcg</td>\n",
       "      <td>ttg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>02</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.SATRL0117008_S16.lab.TRL0117008_S16.iso.1</th>\n",
       "      <td>rpoB</td>\n",
       "      <td>S450L</td>\n",
       "      <td>450.0</td>\n",
       "      <td>450.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tcg</td>\n",
       "      <td>ttg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.SATRL0115669_S18.lab.TRL0115669_S18.iso.1</th>\n",
       "      <td>rpoB</td>\n",
       "      <td>S450L</td>\n",
       "      <td>450.0</td>\n",
       "      <td>450.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tcg</td>\n",
       "      <td>ttg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.XD01227882.lab.XD01227882.iso.1</th>\n",
       "      <td>rpoB</td>\n",
       "      <td>S450L</td>\n",
       "      <td>450.0</td>\n",
       "      <td>450.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tcg</td>\n",
       "      <td>ttg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.WG00269790.lab.WG00269790.iso.1</th>\n",
       "      <td>rpoB</td>\n",
       "      <td>S450L</td>\n",
       "      <td>450.0</td>\n",
       "      <td>450.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tcg</td>\n",
       "      <td>ttg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.YA00166043.lab.YA00166043.iso.1</th>\n",
       "      <td>rpoB</td>\n",
       "      <td>D435V</td>\n",
       "      <td>435.0</td>\n",
       "      <td>435.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gac</td>\n",
       "      <td>gtc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17978 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    GENE MUTATION  POSITION  \\\n",
       "UNIQUEID                                                                      \n",
       "site.02.subj.1033.lab.2013185075.iso.1              rpoB    S450L     450.0   \n",
       "site.02.subj.0739.lab.2013221518.iso.1              rpoB    S450L     450.0   \n",
       "site.02.subj.0604.lab.241032-14.iso.1               rpoB    M587I     587.0   \n",
       "site.02.subj.0104.lab.22A057.iso.1                  rpoB    S450L     450.0   \n",
       "site.02.subj.0885.lab.22A119.iso.1                  rpoB    S450L     450.0   \n",
       "...                                                  ...      ...       ...   \n",
       "site.10.subj.SATRL0117008_S16.lab.TRL0117008_S1...  rpoB    S450L     450.0   \n",
       "site.10.subj.SATRL0115669_S18.lab.TRL0115669_S1...  rpoB    S450L     450.0   \n",
       "site.10.subj.XD01227882.lab.XD01227882.iso.1        rpoB    S450L     450.0   \n",
       "site.10.subj.WG00269790.lab.WG00269790.iso.1        rpoB    S450L     450.0   \n",
       "site.10.subj.YA00166043.lab.YA00166043.iso.1        rpoB    D435V     435.0   \n",
       "\n",
       "                                                    AMINO_ACID_NUMBER  \\\n",
       "UNIQUEID                                                                \n",
       "site.02.subj.1033.lab.2013185075.iso.1                          450.0   \n",
       "site.02.subj.0739.lab.2013221518.iso.1                          450.0   \n",
       "site.02.subj.0604.lab.241032-14.iso.1                           587.0   \n",
       "site.02.subj.0104.lab.22A057.iso.1                              450.0   \n",
       "site.02.subj.0885.lab.22A119.iso.1                              450.0   \n",
       "...                                                               ...   \n",
       "site.10.subj.SATRL0117008_S16.lab.TRL0117008_S1...              450.0   \n",
       "site.10.subj.SATRL0115669_S18.lab.TRL0115669_S1...              450.0   \n",
       "site.10.subj.XD01227882.lab.XD01227882.iso.1                    450.0   \n",
       "site.10.subj.WG00269790.lab.WG00269790.iso.1                    450.0   \n",
       "site.10.subj.YA00166043.lab.YA00166043.iso.1                    435.0   \n",
       "\n",
       "                                                    GENOME_INDEX  \\\n",
       "UNIQUEID                                                           \n",
       "site.02.subj.1033.lab.2013185075.iso.1                       NaN   \n",
       "site.02.subj.0739.lab.2013221518.iso.1                       NaN   \n",
       "site.02.subj.0604.lab.241032-14.iso.1                        NaN   \n",
       "site.02.subj.0104.lab.22A057.iso.1                           NaN   \n",
       "site.02.subj.0885.lab.22A119.iso.1                           NaN   \n",
       "...                                                          ...   \n",
       "site.10.subj.SATRL0117008_S16.lab.TRL0117008_S1...           NaN   \n",
       "site.10.subj.SATRL0115669_S18.lab.TRL0115669_S1...           NaN   \n",
       "site.10.subj.XD01227882.lab.XD01227882.iso.1                 NaN   \n",
       "site.10.subj.WG00269790.lab.WG00269790.iso.1                 NaN   \n",
       "site.10.subj.YA00166043.lab.YA00166043.iso.1                 NaN   \n",
       "\n",
       "                                                    NUCLEOTIDE_NUMBER  REF  \\\n",
       "UNIQUEID                                                                     \n",
       "site.02.subj.1033.lab.2013185075.iso.1                            NaN  tcg   \n",
       "site.02.subj.0739.lab.2013221518.iso.1                            NaN  tcg   \n",
       "site.02.subj.0604.lab.241032-14.iso.1                             NaN  atg   \n",
       "site.02.subj.0104.lab.22A057.iso.1                                NaN  tcg   \n",
       "site.02.subj.0885.lab.22A119.iso.1                                NaN  tcg   \n",
       "...                                                               ...  ...   \n",
       "site.10.subj.SATRL0117008_S16.lab.TRL0117008_S1...                NaN  tcg   \n",
       "site.10.subj.SATRL0115669_S18.lab.TRL0115669_S1...                NaN  tcg   \n",
       "site.10.subj.XD01227882.lab.XD01227882.iso.1                      NaN  tcg   \n",
       "site.10.subj.WG00269790.lab.WG00269790.iso.1                      NaN  tcg   \n",
       "site.10.subj.YA00166043.lab.YA00166043.iso.1                      NaN  gac   \n",
       "\n",
       "                                                    ALT  IS_SNP  IS_INDEL  \\\n",
       "UNIQUEID                                                                    \n",
       "site.02.subj.1033.lab.2013185075.iso.1              ttg    True     False   \n",
       "site.02.subj.0739.lab.2013221518.iso.1              ttg    True     False   \n",
       "site.02.subj.0604.lab.241032-14.iso.1               ata    True     False   \n",
       "site.02.subj.0104.lab.22A057.iso.1                  ttg    True     False   \n",
       "site.02.subj.0885.lab.22A119.iso.1                  ttg    True     False   \n",
       "...                                                 ...     ...       ...   \n",
       "site.10.subj.SATRL0117008_S16.lab.TRL0117008_S1...  ttg    True     False   \n",
       "site.10.subj.SATRL0115669_S18.lab.TRL0115669_S1...  ttg    True     False   \n",
       "site.10.subj.XD01227882.lab.XD01227882.iso.1        ttg    True     False   \n",
       "site.10.subj.WG00269790.lab.WG00269790.iso.1        ttg    True     False   \n",
       "site.10.subj.YA00166043.lab.YA00166043.iso.1        gtc    True     False   \n",
       "\n",
       "                                                    ...  IS_NULL  \\\n",
       "UNIQUEID                                            ...            \n",
       "site.02.subj.1033.lab.2013185075.iso.1              ...    False   \n",
       "site.02.subj.0739.lab.2013221518.iso.1              ...    False   \n",
       "site.02.subj.0604.lab.241032-14.iso.1               ...    False   \n",
       "site.02.subj.0104.lab.22A057.iso.1                  ...    False   \n",
       "site.02.subj.0885.lab.22A119.iso.1                  ...    False   \n",
       "...                                                 ...      ...   \n",
       "site.10.subj.SATRL0117008_S16.lab.TRL0117008_S1...  ...    False   \n",
       "site.10.subj.SATRL0115669_S18.lab.TRL0115669_S1...  ...    False   \n",
       "site.10.subj.XD01227882.lab.XD01227882.iso.1        ...    False   \n",
       "site.10.subj.WG00269790.lab.WG00269790.iso.1        ...    False   \n",
       "site.10.subj.YA00166043.lab.YA00166043.iso.1        ...    False   \n",
       "\n",
       "                                                    IS_FILTER_PASS  \\\n",
       "UNIQUEID                                                             \n",
       "site.02.subj.1033.lab.2013185075.iso.1                        True   \n",
       "site.02.subj.0739.lab.2013221518.iso.1                        True   \n",
       "site.02.subj.0604.lab.241032-14.iso.1                         True   \n",
       "site.02.subj.0104.lab.22A057.iso.1                            True   \n",
       "site.02.subj.0885.lab.22A119.iso.1                            True   \n",
       "...                                                            ...   \n",
       "site.10.subj.SATRL0117008_S16.lab.TRL0117008_S1...            True   \n",
       "site.10.subj.SATRL0115669_S18.lab.TRL0115669_S1...            True   \n",
       "site.10.subj.XD01227882.lab.XD01227882.iso.1                  True   \n",
       "site.10.subj.WG00269790.lab.WG00269790.iso.1                  True   \n",
       "site.10.subj.YA00166043.lab.YA00166043.iso.1                  True   \n",
       "\n",
       "                                                    ELEMENT_TYPE  \\\n",
       "UNIQUEID                                                           \n",
       "site.02.subj.1033.lab.2013185075.iso.1                      GENE   \n",
       "site.02.subj.0739.lab.2013221518.iso.1                      GENE   \n",
       "site.02.subj.0604.lab.241032-14.iso.1                       GENE   \n",
       "site.02.subj.0104.lab.22A057.iso.1                          GENE   \n",
       "site.02.subj.0885.lab.22A119.iso.1                          GENE   \n",
       "...                                                          ...   \n",
       "site.10.subj.SATRL0117008_S16.lab.TRL0117008_S1...          GENE   \n",
       "site.10.subj.SATRL0115669_S18.lab.TRL0115669_S1...          GENE   \n",
       "site.10.subj.XD01227882.lab.XD01227882.iso.1                GENE   \n",
       "site.10.subj.WG00269790.lab.WG00269790.iso.1                GENE   \n",
       "site.10.subj.YA00166043.lab.YA00166043.iso.1                GENE   \n",
       "\n",
       "                                                    MUTATION_TYPE  \\\n",
       "UNIQUEID                                                            \n",
       "site.02.subj.1033.lab.2013185075.iso.1                        AAM   \n",
       "site.02.subj.0739.lab.2013221518.iso.1                        AAM   \n",
       "site.02.subj.0604.lab.241032-14.iso.1                         AAM   \n",
       "site.02.subj.0104.lab.22A057.iso.1                            AAM   \n",
       "site.02.subj.0885.lab.22A119.iso.1                            AAM   \n",
       "...                                                           ...   \n",
       "site.10.subj.SATRL0117008_S16.lab.TRL0117008_S1...            AAM   \n",
       "site.10.subj.SATRL0115669_S18.lab.TRL0115669_S1...            AAM   \n",
       "site.10.subj.XD01227882.lab.XD01227882.iso.1                  AAM   \n",
       "site.10.subj.WG00269790.lab.WG00269790.iso.1                  AAM   \n",
       "site.10.subj.YA00166043.lab.YA00166043.iso.1                  AAM   \n",
       "\n",
       "                                                    INDEL_LENGTH  INDEL_1  \\\n",
       "UNIQUEID                                                                    \n",
       "site.02.subj.1033.lab.2013185075.iso.1                       NaN            \n",
       "site.02.subj.0739.lab.2013221518.iso.1                       NaN            \n",
       "site.02.subj.0604.lab.241032-14.iso.1                        NaN            \n",
       "site.02.subj.0104.lab.22A057.iso.1                           NaN            \n",
       "site.02.subj.0885.lab.22A119.iso.1                           NaN            \n",
       "...                                                          ...      ...   \n",
       "site.10.subj.SATRL0117008_S16.lab.TRL0117008_S1...           NaN            \n",
       "site.10.subj.SATRL0115669_S18.lab.TRL0115669_S1...           NaN            \n",
       "site.10.subj.XD01227882.lab.XD01227882.iso.1                 NaN            \n",
       "site.10.subj.WG00269790.lab.WG00269790.iso.1                 NaN            \n",
       "site.10.subj.YA00166043.lab.YA00166043.iso.1                 NaN            \n",
       "\n",
       "                                                    INDEL_2 SITEID  \\\n",
       "UNIQUEID                                                             \n",
       "site.02.subj.1033.lab.2013185075.iso.1                          02   \n",
       "site.02.subj.0739.lab.2013221518.iso.1                          02   \n",
       "site.02.subj.0604.lab.241032-14.iso.1                           02   \n",
       "site.02.subj.0104.lab.22A057.iso.1                              02   \n",
       "site.02.subj.0885.lab.22A119.iso.1                              02   \n",
       "...                                                     ...    ...   \n",
       "site.10.subj.SATRL0117008_S16.lab.TRL0117008_S1...              10   \n",
       "site.10.subj.SATRL0115669_S18.lab.TRL0115669_S1...              10   \n",
       "site.10.subj.XD01227882.lab.XD01227882.iso.1                    10   \n",
       "site.10.subj.WG00269790.lab.WG00269790.iso.1                    10   \n",
       "site.10.subj.YA00166043.lab.YA00166043.iso.1                    10   \n",
       "\n",
       "                                                   NUMBER_NUCLEOTIDE_CHANGES  \\\n",
       "UNIQUEID                                                                       \n",
       "site.02.subj.1033.lab.2013185075.iso.1                                     1   \n",
       "site.02.subj.0739.lab.2013221518.iso.1                                     1   \n",
       "site.02.subj.0604.lab.241032-14.iso.1                                      1   \n",
       "site.02.subj.0104.lab.22A057.iso.1                                         1   \n",
       "site.02.subj.0885.lab.22A119.iso.1                                         1   \n",
       "...                                                                      ...   \n",
       "site.10.subj.SATRL0117008_S16.lab.TRL0117008_S1...                         1   \n",
       "site.10.subj.SATRL0115669_S18.lab.TRL0115669_S1...                         1   \n",
       "site.10.subj.XD01227882.lab.XD01227882.iso.1                               1   \n",
       "site.10.subj.WG00269790.lab.WG00269790.iso.1                               1   \n",
       "site.10.subj.YA00166043.lab.YA00166043.iso.1                               1   \n",
       "\n",
       "                                                    segid  \n",
       "UNIQUEID                                                   \n",
       "site.02.subj.1033.lab.2013185075.iso.1                  B  \n",
       "site.02.subj.0739.lab.2013221518.iso.1                  B  \n",
       "site.02.subj.0604.lab.241032-14.iso.1                   B  \n",
       "site.02.subj.0104.lab.22A057.iso.1                      B  \n",
       "site.02.subj.0885.lab.22A119.iso.1                      B  \n",
       "...                                                   ...  \n",
       "site.10.subj.SATRL0117008_S16.lab.TRL0117008_S1...      B  \n",
       "site.10.subj.SATRL0115669_S18.lab.TRL0115669_S1...      B  \n",
       "site.10.subj.XD01227882.lab.XD01227882.iso.1            B  \n",
       "site.10.subj.WG00269790.lab.WG00269790.iso.1            B  \n",
       "site.10.subj.YA00166043.lab.YA00166043.iso.1            B  \n",
       "\n",
       "[17978 rows x 25 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#read in CRyPTIC mutation data\n",
    "mutation_df = pd.read_pickle('Data_tables/MUTATIONS-rnap.pkl.gz')\n",
    "mutation_df.drop_duplicates(inplace=True)\n",
    "#remove nulls and ensure mutation is a SNP\n",
    "\n",
    "mutation_df = mutation_df[(~mutation_df['IS_NULL']) & (mutation_df['IS_FILTER_PASS']) & (mutation_df['IS_SNP'])]\n",
    "#ensure mutation is in the coding region\n",
    "mutation_df = mutation_df[mutation_df['IN_CDS']]\n",
    "#ensure mutations are non-synonymous\n",
    "mutation_df = mutation_df[~mutation_df['IS_SYNONYMOUS']]\n",
    "#only investigate mutations in rpoB\n",
    "mutation_df = mutation_df.loc[mutation_df.GENE=='rpoB']\n",
    "#only investigate solo mutations\n",
    "mutation_df = mutation_df.drop_duplicates(subset=['UNIQUEID'], keep=False)\n",
    "#insert segid column\n",
    "mutation_df['segid'] = [i[-1] for i in mutation_df.GENE]\n",
    "mutation_df.set_index('UNIQUEID',verify_integrity=True, inplace=True)\n",
    "mutation_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phenotype data derivation\n",
    "\n",
    "Primary phenotypic data were Minimum Inhibitory Concentrations (MIC) read from bespoke CRyPTIC UKMYC5 and UKMYC6 susceptibility plates on which the same clinical isolates that underwent WGS were plated. Data was pooled form laboraties in 27 countries across 65 continents, and to date, the WHO-endoresed CRYpTIC dataset contains MICs of 15,211 isolates to 13 different anti-TB drugs ([Brankin et al, 2021](#Brankin2021), [Walker et al, 2022](#Walker2022)).\n",
    "   \n",
    "Each UKMYC5/6 96-well microtitre plate contained 13 different 1st and 2nd line anti-TB compounds, including rifampicin, as well as two repurposed drugs (delamanid and bedaquiline) and two positive control wells. Plates were manufactured with the drugs freeze-dried onto each well following a doubling dilution series. After incubation for 14 days, MICs are read by a trained lab technician via either a Sensititre-Vizion Digital MIC viewing system or via a Mirrored Box [(Plate et al, 2018)](#Plate2018), and these are verified by an Automated Mycobacterial Growth Detection Algorithm (AMyGDA) ([Fowler et al, 2018](#Fowler2018); [Plate et al, 2018](#Plate2018)), as well as via a community science platform, BashTheBug. Phenotype readout disagreements between all three methods constituted low-quality phenotypes and were discarded for this study. \n",
    "\n",
    "The dataset contains a resistance vs susceptible binary phenotype calculated directly from MICs (using a critical concentration Epidemiological Cut-off (ECOFF) value of 1.0 mg/L for rifampicin) that acts as the target set for binary classification models ([CRyPTIC, 2019](#CRyPTIC2019))."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG</th>\n",
       "      <th>PHENOTYPE</th>\n",
       "      <th>SOURCE</th>\n",
       "      <th>METHOD_1</th>\n",
       "      <th>METHOD_2</th>\n",
       "      <th>METHOD_3</th>\n",
       "      <th>METHOD_CC</th>\n",
       "      <th>METHOD_MIC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>site.06.subj.06TB_0290.lab.06MIL0881.iso.1</th>\n",
       "      <td>RIF</td>\n",
       "      <td>R</td>\n",
       "      <td>CRyPTIC</td>\n",
       "      <td>liquid media</td>\n",
       "      <td>microdilution plate</td>\n",
       "      <td>UKMYC5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>&gt;4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.PMFR-0724.lab.MFR-199.iso.1</th>\n",
       "      <td>RIF</td>\n",
       "      <td>S</td>\n",
       "      <td>CRyPTIC</td>\n",
       "      <td>liquid media</td>\n",
       "      <td>microdilution plate</td>\n",
       "      <td>UKMYC5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.PTAN-0168.lab.TAN-500.iso.1</th>\n",
       "      <td>RIF</td>\n",
       "      <td>S</td>\n",
       "      <td>CRyPTIC</td>\n",
       "      <td>liquid media</td>\n",
       "      <td>microdilution plate</td>\n",
       "      <td>UKMYC5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>&lt;=0.06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.06.subj.MHL_0185-14.lab.06MIL0212.iso.1</th>\n",
       "      <td>RIF</td>\n",
       "      <td>R</td>\n",
       "      <td>CRyPTIC</td>\n",
       "      <td>liquid media</td>\n",
       "      <td>microdilution plate</td>\n",
       "      <td>UKMYC5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>&gt;4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.06.subj.MHL_1323-15-R.lab.06MIL0060.iso.1</th>\n",
       "      <td>RIF</td>\n",
       "      <td>R</td>\n",
       "      <td>CRyPTIC</td>\n",
       "      <td>liquid media</td>\n",
       "      <td>microdilution plate</td>\n",
       "      <td>UKMYC5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>&gt;4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.PTAN-0340.lab.TAN-347.iso.1</th>\n",
       "      <td>RIF</td>\n",
       "      <td>S</td>\n",
       "      <td>CLIRES</td>\n",
       "      <td>MODS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.PTAN-0252.lab.TAN-578.iso.1</th>\n",
       "      <td>RIF</td>\n",
       "      <td>S</td>\n",
       "      <td>CLIRES</td>\n",
       "      <td>MODS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.PSLM-0841.lab.SLM-108.iso.1</th>\n",
       "      <td>RIF</td>\n",
       "      <td>R</td>\n",
       "      <td>CLIRES</td>\n",
       "      <td>solid media</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.PMK-1021.lab.MK-1825.iso.1</th>\n",
       "      <td>RIF</td>\n",
       "      <td>S</td>\n",
       "      <td>CLIRES</td>\n",
       "      <td>MODS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.00.subj.0000.lab.1.iso.1</th>\n",
       "      <td>RIF</td>\n",
       "      <td>S</td>\n",
       "      <td>CLIRES</td>\n",
       "      <td>liquid media</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>45621 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               DRUG PHENOTYPE   SOURCE  \\\n",
       "UNIQUEID                                                                 \n",
       "site.06.subj.06TB_0290.lab.06MIL0881.iso.1      RIF         R  CRyPTIC   \n",
       "site.05.subj.PMFR-0724.lab.MFR-199.iso.1        RIF         S  CRyPTIC   \n",
       "site.05.subj.PTAN-0168.lab.TAN-500.iso.1        RIF         S  CRyPTIC   \n",
       "site.06.subj.MHL_0185-14.lab.06MIL0212.iso.1    RIF         R  CRyPTIC   \n",
       "site.06.subj.MHL_1323-15-R.lab.06MIL0060.iso.1  RIF         R  CRyPTIC   \n",
       "...                                             ...       ...      ...   \n",
       "site.05.subj.PTAN-0340.lab.TAN-347.iso.1        RIF         S   CLIRES   \n",
       "site.05.subj.PTAN-0252.lab.TAN-578.iso.1        RIF         S   CLIRES   \n",
       "site.05.subj.PSLM-0841.lab.SLM-108.iso.1        RIF         R   CLIRES   \n",
       "site.05.subj.PMK-1021.lab.MK-1825.iso.1         RIF         S   CLIRES   \n",
       "site.00.subj.0000.lab.1.iso.1                   RIF         S   CLIRES   \n",
       "\n",
       "                                                    METHOD_1  \\\n",
       "UNIQUEID                                                       \n",
       "site.06.subj.06TB_0290.lab.06MIL0881.iso.1      liquid media   \n",
       "site.05.subj.PMFR-0724.lab.MFR-199.iso.1        liquid media   \n",
       "site.05.subj.PTAN-0168.lab.TAN-500.iso.1        liquid media   \n",
       "site.06.subj.MHL_0185-14.lab.06MIL0212.iso.1    liquid media   \n",
       "site.06.subj.MHL_1323-15-R.lab.06MIL0060.iso.1  liquid media   \n",
       "...                                                      ...   \n",
       "site.05.subj.PTAN-0340.lab.TAN-347.iso.1                MODS   \n",
       "site.05.subj.PTAN-0252.lab.TAN-578.iso.1                MODS   \n",
       "site.05.subj.PSLM-0841.lab.SLM-108.iso.1         solid media   \n",
       "site.05.subj.PMK-1021.lab.MK-1825.iso.1                 MODS   \n",
       "site.00.subj.0000.lab.1.iso.1                   liquid media   \n",
       "\n",
       "                                                           METHOD_2 METHOD_3  \\\n",
       "UNIQUEID                                                                       \n",
       "site.06.subj.06TB_0290.lab.06MIL0881.iso.1      microdilution plate   UKMYC5   \n",
       "site.05.subj.PMFR-0724.lab.MFR-199.iso.1        microdilution plate   UKMYC5   \n",
       "site.05.subj.PTAN-0168.lab.TAN-500.iso.1        microdilution plate   UKMYC5   \n",
       "site.06.subj.MHL_0185-14.lab.06MIL0212.iso.1    microdilution plate   UKMYC5   \n",
       "site.06.subj.MHL_1323-15-R.lab.06MIL0060.iso.1  microdilution plate   UKMYC5   \n",
       "...                                                             ...      ...   \n",
       "site.05.subj.PTAN-0340.lab.TAN-347.iso.1                       None     None   \n",
       "site.05.subj.PTAN-0252.lab.TAN-578.iso.1                       None     None   \n",
       "site.05.subj.PSLM-0841.lab.SLM-108.iso.1                       None     None   \n",
       "site.05.subj.PMK-1021.lab.MK-1825.iso.1                        None     None   \n",
       "site.00.subj.0000.lab.1.iso.1                                  None     None   \n",
       "\n",
       "                                               METHOD_CC METHOD_MIC  \n",
       "UNIQUEID                                                             \n",
       "site.06.subj.06TB_0290.lab.06MIL0881.iso.1           0.5         >4  \n",
       "site.05.subj.PMFR-0724.lab.MFR-199.iso.1             0.5       0.12  \n",
       "site.05.subj.PTAN-0168.lab.TAN-500.iso.1             0.5     <=0.06  \n",
       "site.06.subj.MHL_0185-14.lab.06MIL0212.iso.1         0.5         >4  \n",
       "site.06.subj.MHL_1323-15-R.lab.06MIL0060.iso.1       0.5         >4  \n",
       "...                                                  ...        ...  \n",
       "site.05.subj.PTAN-0340.lab.TAN-347.iso.1            None       None  \n",
       "site.05.subj.PTAN-0252.lab.TAN-578.iso.1            None       None  \n",
       "site.05.subj.PSLM-0841.lab.SLM-108.iso.1            None       None  \n",
       "site.05.subj.PMK-1021.lab.MK-1825.iso.1             None       None  \n",
       "site.00.subj.0000.lab.1.iso.1                       None       None  \n",
       "\n",
       "[45621 rows x 8 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phenotype_df = pd.read_pickle('Data_tables/DST_MEASUREMENTS-rifamycins.pkl.gz')\n",
    "#filter out rifamycin\n",
    "phenotype_df = phenotype_df.loc[phenotype_df.DRUG=='RIF']\n",
    "phenotype_df.drop_duplicates(subset=[\"UNIQUEID\"], keep=\"first\", inplace=True)\n",
    "#filter out rifamycin\n",
    "phenotype_df.set_index('UNIQUEID',verify_integrity=True, inplace=True)\n",
    "phenotype_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merge phenotype and mutation dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>AMINO_ACID_NUMBER</th>\n",
       "      <th>GENOME_INDEX</th>\n",
       "      <th>NUCLEOTIDE_NUMBER</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IS_INDEL</th>\n",
       "      <th>...</th>\n",
       "      <th>IS_FILTER_PASS</th>\n",
       "      <th>ELEMENT_TYPE</th>\n",
       "      <th>MUTATION_TYPE</th>\n",
       "      <th>INDEL_LENGTH</th>\n",
       "      <th>INDEL_1</th>\n",
       "      <th>INDEL_2</th>\n",
       "      <th>SITEID</th>\n",
       "      <th>NUMBER_NUCLEOTIDE_CHANGES</th>\n",
       "      <th>segid</th>\n",
       "      <th>PHENOTYPE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>site.02.subj.1033.lab.2013185075.iso.1</th>\n",
       "      <td>rpoB</td>\n",
       "      <td>S450L</td>\n",
       "      <td>450.0</td>\n",
       "      <td>450.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tcg</td>\n",
       "      <td>ttg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>02</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.02.subj.0739.lab.2013221518.iso.1</th>\n",
       "      <td>rpoB</td>\n",
       "      <td>S450L</td>\n",
       "      <td>450.0</td>\n",
       "      <td>450.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tcg</td>\n",
       "      <td>ttg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>02</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.02.subj.0604.lab.241032-14.iso.1</th>\n",
       "      <td>rpoB</td>\n",
       "      <td>M587I</td>\n",
       "      <td>587.0</td>\n",
       "      <td>587.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>atg</td>\n",
       "      <td>ata</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>02</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.02.subj.0104.lab.22A057.iso.1</th>\n",
       "      <td>rpoB</td>\n",
       "      <td>S450L</td>\n",
       "      <td>450.0</td>\n",
       "      <td>450.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tcg</td>\n",
       "      <td>ttg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>02</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.02.subj.0885.lab.22A119.iso.1</th>\n",
       "      <td>rpoB</td>\n",
       "      <td>S450L</td>\n",
       "      <td>450.0</td>\n",
       "      <td>450.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tcg</td>\n",
       "      <td>ttg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>02</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.SATRL0116187_S19.lab.TRL0116187_S19.iso.1</th>\n",
       "      <td>rpoB</td>\n",
       "      <td>S450L</td>\n",
       "      <td>450.0</td>\n",
       "      <td>450.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tcg</td>\n",
       "      <td>ttg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.SATRL0117008_S16.lab.TRL0117008_S16.iso.1</th>\n",
       "      <td>rpoB</td>\n",
       "      <td>S450L</td>\n",
       "      <td>450.0</td>\n",
       "      <td>450.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tcg</td>\n",
       "      <td>ttg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.SATRL0115669_S18.lab.TRL0115669_S18.iso.1</th>\n",
       "      <td>rpoB</td>\n",
       "      <td>S450L</td>\n",
       "      <td>450.0</td>\n",
       "      <td>450.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tcg</td>\n",
       "      <td>ttg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.XD01227882.lab.XD01227882.iso.1</th>\n",
       "      <td>rpoB</td>\n",
       "      <td>S450L</td>\n",
       "      <td>450.0</td>\n",
       "      <td>450.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tcg</td>\n",
       "      <td>ttg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.YA00166043.lab.YA00166043.iso.1</th>\n",
       "      <td>rpoB</td>\n",
       "      <td>D435V</td>\n",
       "      <td>435.0</td>\n",
       "      <td>435.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gac</td>\n",
       "      <td>gtc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9874 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    GENE MUTATION  POSITION  \\\n",
       "UNIQUEID                                                                      \n",
       "site.02.subj.1033.lab.2013185075.iso.1              rpoB    S450L     450.0   \n",
       "site.02.subj.0739.lab.2013221518.iso.1              rpoB    S450L     450.0   \n",
       "site.02.subj.0604.lab.241032-14.iso.1               rpoB    M587I     587.0   \n",
       "site.02.subj.0104.lab.22A057.iso.1                  rpoB    S450L     450.0   \n",
       "site.02.subj.0885.lab.22A119.iso.1                  rpoB    S450L     450.0   \n",
       "...                                                  ...      ...       ...   \n",
       "site.10.subj.SATRL0116187_S19.lab.TRL0116187_S1...  rpoB    S450L     450.0   \n",
       "site.10.subj.SATRL0117008_S16.lab.TRL0117008_S1...  rpoB    S450L     450.0   \n",
       "site.10.subj.SATRL0115669_S18.lab.TRL0115669_S1...  rpoB    S450L     450.0   \n",
       "site.10.subj.XD01227882.lab.XD01227882.iso.1        rpoB    S450L     450.0   \n",
       "site.10.subj.YA00166043.lab.YA00166043.iso.1        rpoB    D435V     435.0   \n",
       "\n",
       "                                                    AMINO_ACID_NUMBER  \\\n",
       "UNIQUEID                                                                \n",
       "site.02.subj.1033.lab.2013185075.iso.1                          450.0   \n",
       "site.02.subj.0739.lab.2013221518.iso.1                          450.0   \n",
       "site.02.subj.0604.lab.241032-14.iso.1                           587.0   \n",
       "site.02.subj.0104.lab.22A057.iso.1                              450.0   \n",
       "site.02.subj.0885.lab.22A119.iso.1                              450.0   \n",
       "...                                                               ...   \n",
       "site.10.subj.SATRL0116187_S19.lab.TRL0116187_S1...              450.0   \n",
       "site.10.subj.SATRL0117008_S16.lab.TRL0117008_S1...              450.0   \n",
       "site.10.subj.SATRL0115669_S18.lab.TRL0115669_S1...              450.0   \n",
       "site.10.subj.XD01227882.lab.XD01227882.iso.1                    450.0   \n",
       "site.10.subj.YA00166043.lab.YA00166043.iso.1                    435.0   \n",
       "\n",
       "                                                    GENOME_INDEX  \\\n",
       "UNIQUEID                                                           \n",
       "site.02.subj.1033.lab.2013185075.iso.1                       NaN   \n",
       "site.02.subj.0739.lab.2013221518.iso.1                       NaN   \n",
       "site.02.subj.0604.lab.241032-14.iso.1                        NaN   \n",
       "site.02.subj.0104.lab.22A057.iso.1                           NaN   \n",
       "site.02.subj.0885.lab.22A119.iso.1                           NaN   \n",
       "...                                                          ...   \n",
       "site.10.subj.SATRL0116187_S19.lab.TRL0116187_S1...           NaN   \n",
       "site.10.subj.SATRL0117008_S16.lab.TRL0117008_S1...           NaN   \n",
       "site.10.subj.SATRL0115669_S18.lab.TRL0115669_S1...           NaN   \n",
       "site.10.subj.XD01227882.lab.XD01227882.iso.1                 NaN   \n",
       "site.10.subj.YA00166043.lab.YA00166043.iso.1                 NaN   \n",
       "\n",
       "                                                    NUCLEOTIDE_NUMBER  REF  \\\n",
       "UNIQUEID                                                                     \n",
       "site.02.subj.1033.lab.2013185075.iso.1                            NaN  tcg   \n",
       "site.02.subj.0739.lab.2013221518.iso.1                            NaN  tcg   \n",
       "site.02.subj.0604.lab.241032-14.iso.1                             NaN  atg   \n",
       "site.02.subj.0104.lab.22A057.iso.1                                NaN  tcg   \n",
       "site.02.subj.0885.lab.22A119.iso.1                                NaN  tcg   \n",
       "...                                                               ...  ...   \n",
       "site.10.subj.SATRL0116187_S19.lab.TRL0116187_S1...                NaN  tcg   \n",
       "site.10.subj.SATRL0117008_S16.lab.TRL0117008_S1...                NaN  tcg   \n",
       "site.10.subj.SATRL0115669_S18.lab.TRL0115669_S1...                NaN  tcg   \n",
       "site.10.subj.XD01227882.lab.XD01227882.iso.1                      NaN  tcg   \n",
       "site.10.subj.YA00166043.lab.YA00166043.iso.1                      NaN  gac   \n",
       "\n",
       "                                                    ALT  IS_SNP  IS_INDEL  \\\n",
       "UNIQUEID                                                                    \n",
       "site.02.subj.1033.lab.2013185075.iso.1              ttg    True     False   \n",
       "site.02.subj.0739.lab.2013221518.iso.1              ttg    True     False   \n",
       "site.02.subj.0604.lab.241032-14.iso.1               ata    True     False   \n",
       "site.02.subj.0104.lab.22A057.iso.1                  ttg    True     False   \n",
       "site.02.subj.0885.lab.22A119.iso.1                  ttg    True     False   \n",
       "...                                                 ...     ...       ...   \n",
       "site.10.subj.SATRL0116187_S19.lab.TRL0116187_S1...  ttg    True     False   \n",
       "site.10.subj.SATRL0117008_S16.lab.TRL0117008_S1...  ttg    True     False   \n",
       "site.10.subj.SATRL0115669_S18.lab.TRL0115669_S1...  ttg    True     False   \n",
       "site.10.subj.XD01227882.lab.XD01227882.iso.1        ttg    True     False   \n",
       "site.10.subj.YA00166043.lab.YA00166043.iso.1        gtc    True     False   \n",
       "\n",
       "                                                    ...  IS_FILTER_PASS  \\\n",
       "UNIQUEID                                            ...                   \n",
       "site.02.subj.1033.lab.2013185075.iso.1              ...            True   \n",
       "site.02.subj.0739.lab.2013221518.iso.1              ...            True   \n",
       "site.02.subj.0604.lab.241032-14.iso.1               ...            True   \n",
       "site.02.subj.0104.lab.22A057.iso.1                  ...            True   \n",
       "site.02.subj.0885.lab.22A119.iso.1                  ...            True   \n",
       "...                                                 ...             ...   \n",
       "site.10.subj.SATRL0116187_S19.lab.TRL0116187_S1...  ...            True   \n",
       "site.10.subj.SATRL0117008_S16.lab.TRL0117008_S1...  ...            True   \n",
       "site.10.subj.SATRL0115669_S18.lab.TRL0115669_S1...  ...            True   \n",
       "site.10.subj.XD01227882.lab.XD01227882.iso.1        ...            True   \n",
       "site.10.subj.YA00166043.lab.YA00166043.iso.1        ...            True   \n",
       "\n",
       "                                                    ELEMENT_TYPE  \\\n",
       "UNIQUEID                                                           \n",
       "site.02.subj.1033.lab.2013185075.iso.1                      GENE   \n",
       "site.02.subj.0739.lab.2013221518.iso.1                      GENE   \n",
       "site.02.subj.0604.lab.241032-14.iso.1                       GENE   \n",
       "site.02.subj.0104.lab.22A057.iso.1                          GENE   \n",
       "site.02.subj.0885.lab.22A119.iso.1                          GENE   \n",
       "...                                                          ...   \n",
       "site.10.subj.SATRL0116187_S19.lab.TRL0116187_S1...          GENE   \n",
       "site.10.subj.SATRL0117008_S16.lab.TRL0117008_S1...          GENE   \n",
       "site.10.subj.SATRL0115669_S18.lab.TRL0115669_S1...          GENE   \n",
       "site.10.subj.XD01227882.lab.XD01227882.iso.1                GENE   \n",
       "site.10.subj.YA00166043.lab.YA00166043.iso.1                GENE   \n",
       "\n",
       "                                                    MUTATION_TYPE  \\\n",
       "UNIQUEID                                                            \n",
       "site.02.subj.1033.lab.2013185075.iso.1                        AAM   \n",
       "site.02.subj.0739.lab.2013221518.iso.1                        AAM   \n",
       "site.02.subj.0604.lab.241032-14.iso.1                         AAM   \n",
       "site.02.subj.0104.lab.22A057.iso.1                            AAM   \n",
       "site.02.subj.0885.lab.22A119.iso.1                            AAM   \n",
       "...                                                           ...   \n",
       "site.10.subj.SATRL0116187_S19.lab.TRL0116187_S1...            AAM   \n",
       "site.10.subj.SATRL0117008_S16.lab.TRL0117008_S1...            AAM   \n",
       "site.10.subj.SATRL0115669_S18.lab.TRL0115669_S1...            AAM   \n",
       "site.10.subj.XD01227882.lab.XD01227882.iso.1                  AAM   \n",
       "site.10.subj.YA00166043.lab.YA00166043.iso.1                  AAM   \n",
       "\n",
       "                                                    INDEL_LENGTH  INDEL_1  \\\n",
       "UNIQUEID                                                                    \n",
       "site.02.subj.1033.lab.2013185075.iso.1                       NaN            \n",
       "site.02.subj.0739.lab.2013221518.iso.1                       NaN            \n",
       "site.02.subj.0604.lab.241032-14.iso.1                        NaN            \n",
       "site.02.subj.0104.lab.22A057.iso.1                           NaN            \n",
       "site.02.subj.0885.lab.22A119.iso.1                           NaN            \n",
       "...                                                          ...      ...   \n",
       "site.10.subj.SATRL0116187_S19.lab.TRL0116187_S1...           NaN            \n",
       "site.10.subj.SATRL0117008_S16.lab.TRL0117008_S1...           NaN            \n",
       "site.10.subj.SATRL0115669_S18.lab.TRL0115669_S1...           NaN            \n",
       "site.10.subj.XD01227882.lab.XD01227882.iso.1                 NaN            \n",
       "site.10.subj.YA00166043.lab.YA00166043.iso.1                 NaN            \n",
       "\n",
       "                                                    INDEL_2  SITEID  \\\n",
       "UNIQUEID                                                              \n",
       "site.02.subj.1033.lab.2013185075.iso.1                           02   \n",
       "site.02.subj.0739.lab.2013221518.iso.1                           02   \n",
       "site.02.subj.0604.lab.241032-14.iso.1                            02   \n",
       "site.02.subj.0104.lab.22A057.iso.1                               02   \n",
       "site.02.subj.0885.lab.22A119.iso.1                               02   \n",
       "...                                                     ...     ...   \n",
       "site.10.subj.SATRL0116187_S19.lab.TRL0116187_S1...               10   \n",
       "site.10.subj.SATRL0117008_S16.lab.TRL0117008_S1...               10   \n",
       "site.10.subj.SATRL0115669_S18.lab.TRL0115669_S1...               10   \n",
       "site.10.subj.XD01227882.lab.XD01227882.iso.1                     10   \n",
       "site.10.subj.YA00166043.lab.YA00166043.iso.1                     10   \n",
       "\n",
       "                                                   NUMBER_NUCLEOTIDE_CHANGES  \\\n",
       "UNIQUEID                                                                       \n",
       "site.02.subj.1033.lab.2013185075.iso.1                                     1   \n",
       "site.02.subj.0739.lab.2013221518.iso.1                                     1   \n",
       "site.02.subj.0604.lab.241032-14.iso.1                                      1   \n",
       "site.02.subj.0104.lab.22A057.iso.1                                         1   \n",
       "site.02.subj.0885.lab.22A119.iso.1                                         1   \n",
       "...                                                                      ...   \n",
       "site.10.subj.SATRL0116187_S19.lab.TRL0116187_S1...                         1   \n",
       "site.10.subj.SATRL0117008_S16.lab.TRL0117008_S1...                         1   \n",
       "site.10.subj.SATRL0115669_S18.lab.TRL0115669_S1...                         1   \n",
       "site.10.subj.XD01227882.lab.XD01227882.iso.1                               1   \n",
       "site.10.subj.YA00166043.lab.YA00166043.iso.1                               1   \n",
       "\n",
       "                                                   segid  PHENOTYPE  \n",
       "UNIQUEID                                                             \n",
       "site.02.subj.1033.lab.2013185075.iso.1                 B          R  \n",
       "site.02.subj.0739.lab.2013221518.iso.1                 B          R  \n",
       "site.02.subj.0604.lab.241032-14.iso.1                  B          S  \n",
       "site.02.subj.0104.lab.22A057.iso.1                     B          R  \n",
       "site.02.subj.0885.lab.22A119.iso.1                     B          R  \n",
       "...                                                  ...        ...  \n",
       "site.10.subj.SATRL0116187_S19.lab.TRL0116187_S1...     B          R  \n",
       "site.10.subj.SATRL0117008_S16.lab.TRL0117008_S1...     B          R  \n",
       "site.10.subj.SATRL0115669_S18.lab.TRL0115669_S1...     B          R  \n",
       "site.10.subj.XD01227882.lab.XD01227882.iso.1           B          R  \n",
       "site.10.subj.YA00166043.lab.YA00166043.iso.1           B          R  \n",
       "\n",
       "[9874 rows x 26 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged = mutation_df.join(phenotype_df[['PHENOTYPE']],how='inner')\n",
    "merged = merged.loc[merged.GENE=='rpoB']\n",
    "merged"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This dataframe now contains all isolates with the desired type of mutation in rpoB and corresponding binary resistance phenotype. However, the majority of these are isolates that have the same mutation, particulary S450L as demonstrated below. Rather than building an unbiased model, training on this data would result in heavily weighting the model on the few most commonly observed mutations.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MUTATION</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S450L</th>\n",
       "      <td>6013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D435V</th>\n",
       "      <td>777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H445D</th>\n",
       "      <td>405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H445Y</th>\n",
       "      <td>341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L452P</th>\n",
       "      <td>150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L430P</th>\n",
       "      <td>143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S450W</th>\n",
       "      <td>138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H445L</th>\n",
       "      <td>138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D435Y</th>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I491F</th>\n",
       "      <td>110</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       MUTATION\n",
       "S450L      6013\n",
       "D435V       777\n",
       "H445D       405\n",
       "H445Y       341\n",
       "L452P       150\n",
       "L430P       143\n",
       "S450W       138\n",
       "H445L       138\n",
       "D435Y       119\n",
       "I491F       110"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(merged['MUTATION'].value_counts())[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To account for this, all data was 'compressed' so that rather than training on isolates, the models were trained on unique mutations. A complication, however, arises from the fact many mutations do not consistently result in the same phenotype. To address this, an arbitrary threshold of 0.5 was set, whereby mutations that occur three or more times are classified as their majority phenotype, while mutations that occur at least once are classified as resistant or susceptible if all phenotypes are resistinat or suscpetible."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>PHENOTYPE</th>\n",
       "      <th>segid</th>\n",
       "      <th>I</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "      <th>U</th>\n",
       "      <th>total</th>\n",
       "      <th>prob_R</th>\n",
       "      <th>prob_S</th>\n",
       "      <th>PHENOTYPE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MUTATION</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1002P</th>\n",
       "      <td>B</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A286V</th>\n",
       "      <td>B</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A29T</th>\n",
       "      <td>B</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A29V</th>\n",
       "      <td>B</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A334D</th>\n",
       "      <td>B</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V895F</th>\n",
       "      <td>B</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V946G</th>\n",
       "      <td>B</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y308C</th>\n",
       "      <td>B</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y572C</th>\n",
       "      <td>B</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y651H</th>\n",
       "      <td>B</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>317 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "PHENOTYPE segid  I  R  S  U  total  prob_R  prob_S PHENOTYPE\n",
       "MUTATION                                                    \n",
       "A1002P        B  0  0  3  0      3     0.0     1.0         S\n",
       "A286V         B  0  2  0  0      2     1.0     0.0         R\n",
       "A29T          B  0  0  1  0      1     0.0     1.0         S\n",
       "A29V          B  0  0  1  0      1     0.0     1.0         S\n",
       "A334D         B  0  0  1  0      1     0.0     1.0         S\n",
       "...         ... .. .. .. ..    ...     ...     ...       ...\n",
       "V895F         B  0  0  1  0      1     0.0     1.0         S\n",
       "V946G         B  0  0  1  0      1     0.0     1.0         S\n",
       "Y308C         B  0  1  0  0      1     1.0     0.0         R\n",
       "Y572C         B  0  0  3  0      3     0.0     1.0         S\n",
       "Y651H         B  0  0  1  0      1     0.0     1.0         S\n",
       "\n",
       "[317 rows x 9 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#create a crosstab df to calculate the number (and total) of phenotypes per mutation\n",
    "ct = pd.crosstab([merged.MUTATION, merged.segid], merged.PHENOTYPE)\n",
    "ct['total'] = ct['R'] + ct['S']\n",
    "ct['prob_R'] = ct['R']/ct['total']\n",
    "ct['prob_S'] = ct['S']/ct['total']\n",
    "\n",
    "ct.sort_values(by='total',ascending=False)\n",
    "\n",
    "def calculate_RS(row):\n",
    "    #determine majority phenotype\n",
    "    phenotype='U'\n",
    "    if row['total'] >= 3:\n",
    "        if row['prob_R'] > 0.5:\n",
    "            phenotype='R'\n",
    "        elif row['prob_S'] > 0.5:\n",
    "            phenotype='S'\n",
    "    elif row['total'] >=1:\n",
    "        if row['R'] == row['total']:\n",
    "            phenotype='R'\n",
    "        elif row['S'] == row['total']:\n",
    "            phenotype='S'\n",
    "\n",
    "    return phenotype\n",
    "\n",
    "#determine majority phenotype with calculate_RS()\n",
    "ct['PHENOTYPE'] = ct.apply(calculate_RS, axis=1)\n",
    "#filter for mutations that contain a majority phenotype\n",
    "ct = ct[ct['PHENOTYPE'].isin(['R','S'])]\n",
    "\n",
    "ct.reset_index(inplace=True, level = ['segid'])\n",
    "ct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>SEGID</th>\n",
       "      <th>PHENOTYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1002P</td>\n",
       "      <td>B</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A286V</td>\n",
       "      <td>B</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A29T</td>\n",
       "      <td>B</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A29V</td>\n",
       "      <td>B</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A334D</td>\n",
       "      <td>B</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <td>V895F</td>\n",
       "      <td>B</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>V946G</td>\n",
       "      <td>B</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>314</th>\n",
       "      <td>Y308C</td>\n",
       "      <td>B</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>Y572C</td>\n",
       "      <td>B</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>316</th>\n",
       "      <td>Y651H</td>\n",
       "      <td>B</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>317 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    MUTATION SEGID PHENOTYPE\n",
       "0     A1002P     B         S\n",
       "1      A286V     B         R\n",
       "2       A29T     B         S\n",
       "3       A29V     B         S\n",
       "4      A334D     B         S\n",
       "..       ...   ...       ...\n",
       "312    V895F     B         S\n",
       "313    V946G     B         S\n",
       "314    Y308C     B         R\n",
       "315    Y572C     B         S\n",
       "316    Y651H     B         S\n",
       "\n",
       "[317 rows x 3 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#create the final dataframe, phenotypes for which have been determined by crosstab\n",
    "\n",
    "mutation = [i for i in ct.index]\n",
    "index = [i for i in range(0, len(ct))]\n",
    "segid = [i for i in ct.segid]\n",
    "phenotype = [i for i in ct.PHENOTYPE]\n",
    "df = pd.DataFrame({'MUTATION':mutation, 'SEGID':segid, 'PHENOTYPE':phenotype}, index=index)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Feature Set Derivation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The crystal structure of Mtb RNAP used in this study, as well as the structure from which trajectories were previously built, was recently solved to 3.8Å ([Lin et al, 2017](#Lin2017)). Relevant structural features include a bound DNA-mimicking nucleiv acid scaffold, a synthetic RNA oligomer corresponding to a 2-nt mRNA product, a bound rifampicin molecule, and σA responsible for binding the holoenzyme and modulating transcription initiation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All in-silico structural feature pulldowns were performed using a suite of classes, termed sbmlcore (https://github.com/fowler-lab/sbmlcore); the development of which constituted a large component of the project, and the application of which is demonstrated below. Each class is first instantiated with desired input parameters and the output can be appended to a feature set using a public method within the class. Therefore, a user can add a new strucutral feature to the feature set within 2 lines of code, thereby allowing for a streamlined pipeline with rapid construction of a dataframe ready to be split for machine learning."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Biochemical Differences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Selected mutations-specific biochemical differences include changes in volume, molecular weight, hydropahty and isoelectric point"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bibliography\n",
    "\n",
    "Brankin A, Malone KM, Barilar I, Battaglia S, Pires Brandao A, Maurizio Cabibbe A, Carter J, Maria D, Claxton P, Clifton DA, et al (2021) A data compendium of Mycobacterium tuberculosis 1 antibiotic resistance.<a id='Brankin2021'></a>    \n",
    "\n",
    "CRyPTIC (2021) A generalisable approach to drug susceptibility prediction for M. tuberculosis using machine learning and whole-genome sequencing.<a id='CRyPTIC2021'></a>     \n",
    "\n",
    "CRyPTIC (2019) Epidemiological cutoffs for a 96-well broth microtitre plate for high-throughput research antibiotic susceptibility testing of M.tuberculosis.<a id='CRyPTIC2019'></a>\n",
    "\n",
    "Fowler PW, Cruz ALG, Hoosdally SJ, Jarrett L, Borroni E, Chiacchiaretta M, Rathod P, Lehmann S, Molodtsov N, Walker TM, et al (2018) Automated detection of bacterial growth on 96-well plates for high-throughput drug susceptibility testing of mycobacterium tuberculosis. Microbiology (United Kingdom) 164: 1522–1530<a id=Fowler2018></a>\n",
    "\n",
    "Lin W, Mandal S, Degen D, Liu Y, Ebright YW, Li S, Feng Y, Zhang Y, Mandal S, Jiang Y, et al (2017) Structural Basis of Mycobacterium tuberculosis Transcription and Transcription Inhibition. Molecular Cell 66: 169-179.e8<a id=Lin2017></a>\n",
    "\n",
    "Mvelase NR, Pillay M, Sibanda W, Ngozo JN, Brust JCM & Mlisana KP (2019) RpoB Mutations Causing Discordant Rifampicin Susceptibility in Mycobacterium tuberculosis: Retrospective analysis of prevalence, phenotypic, genotypic, and treatment outcomes. Open Forum Infectious Diseases 6 <a id='Mvelase2019'></a>\n",
    "\n",
    "Plate M, Bedaquiline C, Walker TM, Grazian C, Davies TJ, Peto TEA, Crook DW, Fowler PW & Cirillo DM (2018) Validating a 14-Drug Microtiter Plate Containing Bedaquiline and Delamanid for Large-Scale Research Susceptibility Testing of Mycobacterium tuberculosis. 62: 1–15<a id='Plate2018'></a>      \n",
    "\n",
    "Walker TM, Fowler PW, Knaggs J, Hunt M, Peto TE, Walker AS, Crook DW, Walker TM, Miotto P, Cirillo DM, et al (2022) The 2021 WHO catalogue of Mycobacterium tuberculosis complex mutations associated with drug resistance: a genotypic analysis. The Lancet Microbe 3: e265–e273<a id='Walker2022'></a>   \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[CRyPTIC, 2021](#CRyPTIC,2021)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "e86185a29b880bbe385911a49b7671122849e75b1e6a606f7ab424b543a1e33a"
  },
  "kernelspec": {
   "display_name": "Python 3.10.4",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
